x-pack/legacy/plugins/apm/public/components/app/ServiceMap/index.tsx

```diff
@@ -37,37 +54,159 @@ ${theme.euiColorLightShade}`,
   margin: `-${theme.gutterTypes.gutterLarge}`
 };
 
+const MAX_REQUESTS = 5;
+
 export function ServiceMap({ serviceName }: ServiceMapProps) {
-  const {
-    urlParams: { start, end }
-  } = useUrlParams();
+  const callApmApi = useCallApmApi();
+  const license = useLicense();
+  const { search } = useLocation();
+  const { urlParams, uiFilters } = useUrlParams();
+  const { notifications } = useApmPluginContext().core;
+  const params = useDeepObjectIdentity({
+    start: urlParams.start,
+    end: urlParams.end,
+    environment: urlParams.environment,
+    serviceName,
+    uiFilters: {
+      ...uiFilters,
+      environment: undefined
+    }
+  });
+
+  const renderedElements = useRef<ElementDefinition[]>([]);
+  const openToast = useRef<string | null>(null);
+
+  const [responses, setResponses] = useState<ServiceMapAPIResponse[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [percentageLoaded, setPercentageLoaded] = useState(0);
+  const [, _setUnusedState] = useState(false);
+
+  const elements = useMemo(() => getCytoscapeElements(responses, search), [
+    responses,
+    search
+  ]);
+
+  const forceUpdate = useCallback(() => _setUnusedState(value => !value), []);
+
+  const getNext = useCallback(
+    async (input: { reset?: boolean; after?: string | undefined }) => {
+      const { start, end, uiFilters: strippedUiFilters, ...query } = params;
+
+      if (input.reset) {
+        renderedElements.current = [];
+        setResponses([]);
+      }
 
-  const { data } = useFetcher(
-    callApmApi => {
       if (start && end) {
-        return callApmApi({
-          pathname: '/api/apm/service-map',
-          params: { query: { start, end } }
-        });
+        setIsLoading(true);
```

Comment:
The loading state is already updating accordingly in `callApmApi`. Why the custom handling?

Reply:
Correction: loading state is automatically handled if you use `useFetcher` instead of `callApmApi`. My guess is that `callApmApi` was chosen to make the fetch conditional. I have a gut feeling that this leads to procedural code with more edge cases. Instead I'd suggest using `useFetcher` and handle the conditional loading in the callback. This way we might be able to get rid of `getNext` since the `useFetcher` will automatically be invoked on every render